hathor/indexes/memory_deps_index.py

```diff
@@ -0,0 +1,139 @@
+                continue
+            rev_deps = self._rev_dep_index[dep]
+            if tx.hash in rev_deps:
+                rev_deps.remove(tx.hash)
+            if not rev_deps:
+                del self._rev_dep_index[dep]
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        """ Removes from ready for validation set.
+        """
+        self._txs_with_deps_ready.discard(tx)
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
+        if dry_run:
+            cur_ready = self._txs_with_deps_ready.copy()
+        else:
+            cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+        while cur_ready:
+            yield from sorted(cur_ready, key=lambda tx_hash: tx_storage.get_transaction(tx_hash).timestamp)
+            if dry_run:
+                cur_ready = self._txs_with_deps_ready - cur_ready
+            else:
+                cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+
+    def iter(self) -> Iterator[bytes]:
+        yield from self._rev_dep_index.keys()
+
+    def _get_rev_deps(self, tx: bytes) -> FrozenSet[bytes]:
+        """Get all txs that depend on the given tx (i.e. its reverse depdendencies)."""
+        return frozenset(self._rev_dep_index.get(tx, set()))
+
+    def known_children(self, tx: BaseTransaction) -> List[bytes]:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        it_rev_deps = map(tx.storage.get_transaction, self._get_rev_deps(tx.hash))
+        return [not_none(rev.hash) for rev in it_rev_deps if tx.hash in rev.parents]
+
+    # needed-txs-index methods:
+
+    def has_needed_tx(self) -> bool:
+        return bool(self._needed_txs_index)
+
+    def is_tx_needed(self, tx: bytes) -> bool:
+        return tx in self._needed_txs_index
+
+    def remove_from_needed_index(self, tx: bytes) -> None:
+        self._needed_txs_index.pop(tx, None)
+
+    def get_next_needed_tx(self) -> bytes:
+        # This strategy maximizes the chance to download multiple txs on the same stream
+        # find the tx with highest "height"
+        # XXX: we could cache this onto `needed_txs` so we don't have to fetch txs every time
+        height, start_hash, tx = max((h, s, t) for t, (h, s) in self._needed_txs_index.items())
+        self.log.debug('next needed tx start', needed=len(self._needed_txs_index), start=start_hash.hex(),
+                       height=height, needed_tx=tx.hex())
+        return start_hash
+
+    def _add_needed(self, tx: BaseTransaction) -> None:
+        """This method is idempotent, because self.update needs it to be indempotent."""
+        assert tx.storage is not None
+        tx_storage = tx.storage
+
+        height = get_requested_from_height(tx)
+        self.log.debug('add needed deps', tx=tx.hash_hex, height=height, type=type(tx).__name__)
+        for tx_hash in tx.get_all_dependencies():
```

Comment:
In the previous implementation, we were iterating only through the parents here. What changed?

Reply:
Although using the parents is sufficient, using all the dependencies leads to a slightly better order and should also reach a known tx sooner (because this only makes sense for sync-v2).

Reply:
I just remembered that because in sync-v2 we don't sync voided blocks, I think there can be cases where a tx is only reachable if you also follow the inputs (that is, using dependencies instead of only parents).